WalletWasabi.Gui/Tabs/EncryptionManager/DecryptMessageViewModel.cs

```diff
@@ -0,0 +1,221 @@
+			set => this.RaiseAndSetIfChanged(ref _addresses, value);
+		}
+
+		public ReactiveCommand DecryptCommand { get; }
+		public ReactiveCommand MyPublicKeyCommand { get; }
+
+		public EncryptionManagerViewModel Owner { get; }
+
+		public DecryptMessageViewModel(EncryptionManagerViewModel owner) : base("Decrypt Message")
+		{
+			Owner = owner;
+
+			var canDecrypt = this.WhenAnyValue(x => x.EncryptedMessage,
+				(message) =>
+					!string.IsNullOrEmpty(message));
+
+			DecryptCommand = ReactiveCommand.Create(
+				() =>
+				{
+					WarningMessage = "";
+					DecryptedMessage = DecryptMessage(EncryptedMessage, MyPublicKey, Password);
+				},
+				canDecrypt
+			);
+
+			DecryptCommand.ThrownExceptions.Subscribe(ex =>
+			{
+				ShowWarnMessage(ex.Message);
+			});
+
+			MyPublicKeyCommand = ReactiveCommand.Create(
+				() =>
+				{
+					WarningMessage = "";
+					DecryptedMessage = DecryptMessage(EncryptedMessage, MyPublicKey, Password);
+				},
+				canDecrypt
+			);
+
+			MyPublicKeyCommand.ThrownExceptions.Subscribe(ex =>
+			{
+				ShowWarnMessage(ex.Message);
+			});
+
+			this.WhenAnyValue(x => x.SelectedItem)
+				.Subscribe((x) => MyPublicKey = x is null ? "" : x.PubKey);
+
+			this.WhenAnyValue(x => x.AddressSearch)
+				.Throttle(TimeSpan.FromMilliseconds(500)).Subscribe((searchText) =>
+			   {
+				   if (string.IsNullOrEmpty(searchText))
+					   Addresses = new ObservableCollection<AddressPubKeyViewModel>(_allKeysViewModels);
+				   else
+				   {
+					   Addresses = new ObservableCollection<AddressPubKeyViewModel>(_allKeysViewModels.Where(vm => vm.Address.Contains(searchText)));
+					   if (Addresses.Count == 1)
+					   {
+						   SelectedItem = Addresses.First();
+					   }
+
+					   IsSearchResultExpanded = Addresses.Any();
+				   }
+			   });
+
+			IEnumerable<HdPubKey> keys = Global.WalletService.KeyManager.GetKeys();
+			_allKeysViewModels = keys.Select(a => new AddressPubKeyViewModel(a));
+		}
+
+		private void ShowWarnMessage(string message)
+		{
+			_cancellationTokenSource.Cancel();
+			_cancellationTokenSource.Dispose();
+			_cancellationTokenSource = new CancellationTokenSource();
+			WarningMessage = message;
+			Dispatcher.UIThread.Post(async () =>
+			{
+				try
+				{
+					await Task.Delay(7000, _cancellationTokenSource.Token);
+					WarningMessage = "";
+				}
+				catch (Exception) { };
+			});
+		}
+
+		private string DecryptMessage(string message, string pubKeyHex, string password)
+		{
+			password = Guard.Correct(password);
+			PubKey pk = new PubKey(pubKeyHex);
+			HdPubKey hdPubKey = Global.WalletService.KeyManager.GetKeys().FirstOrDefault(k => k.PubKey == pk);
+			if (hdPubKey == null)
+			{
+				throw new InvalidOperationException("Could not find the corresponding address in your wallet for that public key.");
+			}
+
+			(ExtKey secret, HdPubKey pubKey) secret = Global.WalletService.KeyManager.GetSecretsAndPubKeyPairs(password, hdPubKey.PubKey.ScriptPubKey).FirstOrDefault();
```

Comment:
I would prefer call the variable `extKey` instead of `secret`.

Reply:
fixed